src/query/interpret/eval.hpp

```diff
@@ -236,48 +237,39 @@ class ExpressionEvaluator : public ExpressionVisitor<TypedValue> {
       return {};
     };
 
-    auto get_cached_id = [&in_list]() -> std::string {
-      std::string cached_id;
-      if (in_list.expression2_->GetTypeInfo() == ListLiteral::kType) {
-        std::stringstream ss;
-        ss << static_cast<const void *>(in_list.expression2_);
-        cached_id = ss.str();
-        spdlog::trace("ListLiteral in InListOperator: {}", cached_id);
-      } else if (in_list.expression2_->GetTypeInfo() == Identifier::kType) {
-        auto *identifier = utils::Downcast<Identifier>(in_list.expression2_);
-        cached_id = identifier->name_;
-        spdlog::trace("Identifier in InListOperator: {}", cached_id);
-      } else {
-        spdlog::trace("InListOperator got {} as expression2_", in_list.expression2_->GetTypeInfo().name);
-      }
-      return cached_id;
-    };
-    // These checks need to be done in both cases, whether we are caching or not
-    get_list_literal();
-    auto preoperational_checks = do_list_literal_checks();
-    if (preoperational_checks) {
-      return std::move(*preoperational_checks);
-    }
-    auto cached_id = get_cached_id();
+    const auto cached_id = memgraph::utils::GetFrameChangeId(in_list);
 
-    const auto do_cache{frame_change_collector_ != nullptr &&
-                        frame_change_collector_->ContainsTrackingValue(cached_id)};
+    const auto do_cache{frame_change_collector_ != nullptr && cached_id &&
+                        frame_change_collector_->IsKeyTracked(*cached_id)};
     if (do_cache) {
-      if (!frame_change_collector_->IsTrackingValueCached(cached_id)) {
-        auto &cached_value = frame_change_collector_->GetCachedValue(cached_id);
-        cached_value.cache_value_(cached_value, *_list_ptr);
+      if (!frame_change_collector_->IsKeyValueCached(*cached_id)) {
+        get_list_literal();
```

Comment:
This is done in if only once when value is cached. Afterwards if value is cached it doesn't make sense to do it. If I put it before and do it every time, then we lose on performance because we are evaluating this expression (list literal or getting value from frame in case of identifier) and doing it every time.

Reply:
But why do you do it again after the if do_cache?